src/alloc.c

```diff
@@ -414,18 +414,18 @@ static void unmark_symbols_(jl_sym_t *root)
 
 void jl_unmark_symbols(void) { unmark_symbols_(symtab); }
 
-static jl_sym_t **symtab_lookup(jl_sym_t **ptree, const char *str, jl_sym_t **parent)
+static jl_sym_t **symtab_lookup(jl_sym_t **ptree, const char *str, size_t len, jl_sym_t **parent)
 {
     int x;
     if (parent != NULL) *parent = NULL;
-    uptrint_t h = hash_symbol(str, strlen(str));
+    uptrint_t h = hash_symbol(str, len);
 
     // Tree nodes sorted by major key of (int(hash)) and minor key o (str).
     while (*ptree != NULL) {
         x = (int)(h-(*ptree)->hash);
         if (x == 0) {
-            x = strcmp(str, (*ptree)->name);
-            if (x == 0)
+            x = strncmp(str, (*ptree)->name, len);
```

Comment:
since they are constructed to be null terminated with no embedded null characters, it likely could be `memcmp` (it'll also stop at the first difference), but this should be the same